asteroid/engine/gan_system.py

```diff
@@ -0,0 +1,172 @@
+from collections import OrderedDict
+import soundfile as sf
+import pytorch_lightning as pl
+import torch
+import os
+from torch_stoi import NegSTOILoss
+from asteroid.losses import PITLossWrapper
+
+
+class GanSystem(pl.LightningModule):
+
+    """ Base class for GAN training systems.
+       Contains a generator a discriminator, two optimizers,
+       a generator loss function, a discriminator loss function,
+       a validation loss function,
+       training and validation dataloaders and learning rate scheduler.
+
+       Args:
+           discriminator (torch.nn.Module): Instance of discriminator (d).
+           generator (torch.nn.Module): Instance of generator (g).
+           opt_d (torch.optim.Optimizer): Instance of optimizer for d.
+           opt_g (torch.optim.Optimizer): Instance of optimizer for g.
+           discriminator_loss (callable): Loss function with signature
+               (est_targets, targets).
+            generator_loss (callable): Loss function with signature
+               (est_targets, targets).
+           train_loader (torch.utils.data.DataLoader): Training dataloader.
+           val_loader (torch.utils.data.DataLoader): Validation dataloader.
+           scheduler_d (torch.optim.lr_scheduler._LRScheduler): Instance
+               of learning rate schedulers for d.
+           scheduler_g (torch.optim.lr_scheduler._LRScheduler): Instance
+               of learning rate schedulers for g.
+           conf: Anything to be saved with the checkpoints during training.
+               The config dictionary to re-instantiate the run for example.
+       """
+
+    def __init__(self, discriminator, generator, opt_d, opt_g,
+                 discriminator_loss, generator_loss, train_loader,
+                 validation_loss=PITLossWrapper(NegSTOILoss(), pit_from='pw_pt'),
+                 val_loader=None, scheduler_d=None, scheduler_g=None,
+                 conf=None):
+
+        super(GanSystem, self).__init__()
+        self.discriminator = discriminator
+        self.generator = generator
+        self.opt_d = opt_d
+        self.opt_g = opt_g
+        self.d_loss = discriminator_loss
+        self.g_loss = generator_loss
+        self.validation_loss = validation_loss
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.scheduler_d = scheduler_d
+        self.scheduler_g = scheduler_g
+        self.conf = conf
+        self.estimates = None
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def training_step(self, batch, batch_nb, optimizer_idx):
+        # Get data from data_loader
+        inputs, targets = batch
+        # Forward inputs
+        self.estimates = self(inputs)
+        # Train discriminator
+        if optimizer_idx == 0:
+            # Compute D loss for targets
+            est_true_labels = self.discriminator(targets, inputs, targets)
```

Comment:
Is this discriminator call completely generic?

Reply:
I think so, we basically give everything to the network in case someone need the true ref or the noisy ref

Reply:
But this is written nowhere, so the user has to come here to find what the `GanSystem` expects of his network. If he comes this far to read the code, he can re-implement his own System. 
I'm not sure this will be useful. It will either require huge amount of docs or be too simplistic. 
And the code is not so because most args are positional.. 
Maybe it's useful to have the backbone with the optimizer, the scheduler etc.. so that it doesn't have to be rewritten. However, the training step should either be super simple, or `NotImplementedError` IMO. @popcornell any thoughts on that? You know GAN way better than me.

Reply:
why it does take 3 arguments with two arguments identical ? Plain discriminator should have one argument (one time is the enhanced signal another time is the true clean signal). Are you using relativistic discriminator ?

Reply:
> Maybe it's useful to have the backbone with the optimizer, the scheduler etc.. so that it doesn't have to be rewritten. However, the training step should either be super simple, or NotImplementedError IMO. @popcornell any thoughts on that? You know GAN way better than me.

GANs are quite sensitive and difficult to train.... plus there are tons of different types some differs only for the output activation of discriminator and loss function but others (like relativistic discriminator ) requires changes to the training loop.  Also because are hard to train one usually logs everything to tensorboard in the training loop. I mean it is nice to have a GAN example and maybe a GAN system. But for real research it will be bypassed probably.

Reply:
I will actually like to do some barebone system which bypasses most of lightning functionality (also grad accumulation). For example one could wish to have different virtual batches for generator and discriminator ( now with wasserstein gan and such is less needed but still...) . Also useful for other techniques which require full control over the training loop.

Reply:
Thanks for having a look !

Reply:
> why it does take 3 arguments with two arguments identical ? Plain discriminator should have one argument (one time is the enhanced signal another time is the true clean signal). Are you using relativistic discriminator ?

Sometimes the discriminator requires more than just the enhanced signal or the true clean signal. Some discriminator actually compare the enhanced signal to the true Signal ( ex MetricGAN)  some discriminator use the enhanced or the clean signal + the noisy input (SEGAN). This is why I made the discriminator have the args.

Reply:
For the current use cases, this choice can make sense. But the audio GAN landscape is so broad that we cannot aim at covering it all..